File: src/main/java/me/anton/sickcore/api/utils/minecraft/bukkit/item/ItemBuilder.java
package me.anton.sickcore.api.utils.minecraft.bukkit.item;

import lombok.Getter;
import me.anton.sickcore.api.player.bukkitPlayer.BukkitPlayer;
import me.anton.sickcore.api.utils.minecraft.bukkit.inventory.InventoryBuilder;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.Color;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.LeatherArmorMeta;
import org.bukkit.inventory.meta.SkullMeta;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;
import java.util.function.Consumer;

@Getter
public class ItemBuilder {

    @Getter
    private static HashMap<UUID, HashMap<ItemStack, ItemBuilder>> playerHandler = new HashMap<>();

    private ItemStack itemStack;
    private ItemMeta itemMeta;
    private Consumer<PlayerInteractEvent> itemClickEvent;
    private final BukkitPlayer player;
    private InventoryBuilder builder;

    public ItemBuilder(Material material, @Nullable BukkitPlayer player){
        this.itemStack = new ItemStack(material, 1);
        this.itemMeta = itemStack.getItemMeta();
        this.player = player;
        this.itemClickEvent = null;
    }

    public ItemBuilder(Material material, int amount,@Nullable BukkitPlayer player){
        this.itemStack = new ItemStack(material, amount);
        this.itemMeta = itemStack.getItemMeta();
        this.player = player;
        this.itemClickEvent = null;
    }

    public ItemBuilder(ItemStack itemStack,@Nullable BukkitPlayer player){
        this.itemStack = itemStack;
        this.itemMeta = itemStack.getItemMeta();
        this.player = player;
        this.itemClickEvent = null;
    }

    public ItemBuilder addEnchantment(Enchantment enchantment, int level){
        itemMeta.addEnchant(enchantment, level, true);
        return this;
    }

    public ItemBuilder setEnchanted(boolean value){
        if (!value){
            itemMeta.getEnchants().forEach((enchantment, id) -> itemMeta.removeEnchant(enchantment));
            itemMeta.removeItemFlags(ItemFlag.HIDE_ENCHANTS);
            return this;
        }
        itemMeta.addEnchant(Enchantment.MENDING, 1, true);
        itemMeta.addItemFlags(ItemFlag.HIDE_ENCHANTS);
        return this;
    }

    public ItemBuilder setName(@Nullable String title){
        if (title == null)title = "§6";
        itemMeta.displayName(Component.text(title));
        return this;
    }

    public ItemBuilder addItemFlags(@NotNull ItemFlag... itemFlags){
        itemMeta.addItemFlags(itemFlags);
        return this;
    }

    public ItemBuilder setLore(@NotNull String... lore){
        List<Component> lores = new ArrayList();
        for (String loreline : lore)lores.add(Component.text(loreline));
        itemMeta.lore(lores);
        return this;
    }

    public ItemBuilder setLore(@NotNull List<String> lore){
        List<Component> lores = new ArrayList();
        lore.forEach(loreline -> lores.add(Component.text(loreline)));
        itemMeta.lore(lores);
        return this;
    }
    public ItemBuilder setAmount(int amount){
        itemStack.setAmount(amount);
        return this;
    }

    public void onPlayerClick(PlayerInteractEvent event){
        if (itemClickEvent == null)return;
        this.itemClickEvent.accept(event);
    }

    public ItemBuilder setPlayerEvent(Consumer<PlayerInteractEvent> event){
        HashMap<ItemStack, ItemBuilder> add;
        if (playerHandler.containsKey(player.getUniqueID()))add = new HashMap<>(playerHandler.get(player.getUniqueID()));
        else add = new HashMap<>();
        add.put(this.build(), this);
        playerHandler.put(player.api().getUUID(), add);
        this.itemClickEvent = event;
        return this;
    }

    public ItemBuilder setLeatherArmorColor(@NotNull Color color){
        LeatherArmorMeta meta = (LeatherArmorMeta) itemMeta;
        meta.setColor(color);
        return this;
    }

    public ItemBuilder setDurability(int damage){
        Damageable damageable = (Damageable) itemMeta;
        damageable.setDamage(damage);
        return this;
    }

    public ItemBuilder setUnbreakable(boolean value){
        itemMeta.setUnbreakable(value);
        return this;
    }

    public ItemBuilder setSkull(UUID owner){
        if (!(itemStack.getType() == Material.PLAYER_HEAD)) return this;
        SkullMeta skullMeta = (SkullMeta) itemMeta;
        skullMeta.setOwningPlayer(Bukkit.getOfflinePlayer(owner));
        return this;
    }

    public ItemStack build(){
        this.itemStack.setItemMeta(this.itemMeta);
        return this.itemStack;
    }

}

File: src/main/java/me/anton/sickcore/utils/paper/ItemBuilder.kt
package me.anton.sickcore.utils.paper

import com.mojang.authlib.GameProfile
import com.mojang.authlib.properties.Property
import kotlinx.coroutines.launch
import me.anton.sickcore.core.Core
import me.anton.sickcore.core.player.SickPlayer
import me.anton.sickcore.core.player.SickPlayers
import net.axay.kspigot.event.listen
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Bukkit
import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Damageable
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.inventory.meta.LeatherArmorMeta
import org.bukkit.inventory.meta.SkullMeta
import org.bukkit.persistence.PersistentDataType
import org.litote.kmongo.bson
import java.lang.reflect.Field
import java.util.Base64
import java.util.UUID

class ItemBuilder(val player: SickPlayer, val material: Material) {

    var clickEvent: ((PlayerInteractEvent) -> Unit)? = null
    var itemStack = ItemStack(material)
    var itemMeta: ItemMeta = itemStack.itemMeta

    init {
        ItemBuilders.addItem(this)
    }

    fun setName(component: Component?): ItemBuilder{
        if (component == null){
            itemMeta.displayName(Component.text("§7"))
            return this
        }
        itemMeta.displayName(component)
        return this
    }

    fun addUnformattedLore(vararg component: Component): ItemBuilder{
        itemMeta.lore(component.asList())

        return this
    }

    fun addLore(vararg component: Component): ItemBuilder{
        val lores = ArrayList<Component>()
        component.forEach {
            lores.add(MiniMessage.miniMessage().deserialize("<italic> ${MiniMessage.miniMessage().serialize(it)} </italic>"))
        }

        itemMeta.lore(lores.toMutableList())
        return this
    }

    fun addItemFlag(vararg flag: ItemFlag):ItemBuilder{
        flag.forEach { itemMeta.addItemFlags(it) }
        return this
    }

    fun glowing(value: Boolean): ItemBuilder{
        addEnchantment(Enchantment.ARROW_FIRE, 1)
        addItemFlag(ItemFlag.HIDE_ENCHANTS)
        return this
    }

    fun addEnchantment(enchantment: Enchantment, level: Int):ItemBuilder{
        itemMeta.addEnchant(enchantment, level, true)
        return this
    }

    fun setAmount(amount: Int): ItemBuilder{
        itemStack.amount = amount
        return this
    }

    fun asClickableItem(interaction: (PlayerInteractEvent) -> Unit): ItemBuilder{
        clickEvent = interaction
        return this
    }

    fun setDurability(damage: Int): ItemBuilder{
        val damageable = itemMeta as Damageable
        damageable.damage(damage.toDouble())
        return this
    }

    fun setUnbreakable(value: Boolean): ItemBuilder{
        itemMeta.isUnbreakable = value
        return this
    }

    fun setSkull(owner: UUID): ItemBuilder{
        if (itemStack.type != Material.PLAYER_HEAD)return this
        val meta = itemMeta as SkullMeta
        meta.owningPlayer = Bukkit.getOfflinePlayer(owner)
        return this
    }

    fun setSkull(base64: String): ItemBuilder{
        if (itemStack.type != Material.PLAYER_HEAD)return this
        val profile = GameProfile(UUID.randomUUID(), null)
        val encoded = Base64.getEncoder().encode(String.format("{textures:{SKIN:{url:\"%s\"}}}", base64).toByteArray())
        profile.properties.put("textures", Property("textures", String(encoded)))
        var field:Field? = null
        val skullMeta = itemMeta as SkullMeta

        field = skullMeta.javaClass.getDeclaredField("profile")

        assert(field != null)
        field.isAccessible = true
        field.set(skullMeta, profile)

        return this
    }

    fun setLeatherArmorColor(color: Color): ItemBuilder{
        val meta = itemMeta as LeatherArmorMeta
        meta.setColor(color)
        return this
    }

    fun addNameSpacedKey(key: NamespacedKey, type: PersistentDataType<Any, Any>, value: Any):ItemBuilder{
        itemMeta.persistentDataContainer.set(key, type, value)

        return this
    }

    fun build(): ItemStack{
        itemStack.itemMeta = itemMeta

        return itemStack
    }

}

class ItemBuilders{
    companion object{
        val items = HashMap<SickPlayer, ArrayList<ItemBuilder>>()

        fun addItem(builder: ItemBuilder){
            val builders = items[builder.player]
            builders?.add(builder)
            items.replace(builder.player, builders!!)
        }
    }

    fun registerBuilders(){
        listen<PlayerInteractEvent> {
            Core.instance.databaseScope.launch {
                if (!it.hasItem())return@launch
                if (!items.containsKey(SickPlayers.getSickPlayer(it.player.uniqueId)))return@launch
                val items = items[SickPlayers.getSickPlayer(it.player.uniqueId)]
                val itemStacks = ArrayList<ItemStack>()
                items?.forEach { itemStacks.add(it.itemStack) }
                if (!itemStacks.contains(it.item))return@launch
                var item: ItemBuilder? = null
                items?.forEach{builder ->
                    if (builder.itemStack == it.item)item = builder
                }
                item?.clickEvent?.invoke(it)
            }
        }
    }
}